File: server/api/schemas/responses/ui_prefs.py
"""UI-preference envelopes and PUT bodies (OpenAPI SoT for web/src/api/uiPrefs)."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from server.worksets_const import SYSTEM_WORKSET_ID

BoardGanttViewMode = Literal["day", "month"]
TimelineEventStatus = Literal["pending", "confirmed", "completed"]
AssistantMessageRole = Literal["user", "assistant"]
NotifyHistoryStatus = Literal["success", "failure"]
SpacePttMode = Literal["hold", "toggle"]


class SourceFilterSelectionSchema(BaseModel):
    """Hierarchical source multi-select (``null`` at parent = all sources)."""

    model_config = ConfigDict(extra="forbid")

    taskIds: list[str] = Field(default_factory=list)
    worksetIds: list[str] = Field(default_factory=list)


class BoardMapViewSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    center: list[float] = Field(min_length=2, max_length=2)
    zoom: float


class BoardWidgetSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    i: str
    type: str
    col: int
    row: int
    sizeId: str
    z: int | None = None


class BoardLayoutSchema(BaseModel):
    """Layout blob stored under ``ops_board_layout`` (v16 widgets mosaic)."""

    model_config = ConfigDict(extra="forbid")

    version: int
    widgets: list[BoardWidgetSchema]


class BoardWidgetStateSchema(BaseModel):
    """Per-widget map / source-filter / gantt zoom state."""

    model_config = ConfigDict(extra="forbid")

    mapViews: dict[str, BoardMapViewSchema]
    sourceFilters: dict[str, SourceFilterSelectionSchema | None]
    ganttViewModes: dict[str, BoardGanttViewMode]


class BoardPrefsResponse(BaseModel):
    configured: bool
    layout: BoardLayoutSchema | None = None
    widgetState: BoardWidgetStateSchema | None = None


class BoardPrefsPutBody(BaseModel):
    """Partial board upsert. Omit a key to leave unchanged; ``null`` clears it."""

    model_config = ConfigDict(extra="forbid")

    layout: BoardLayoutSchema | None = None
    widgetState: BoardWidgetStateSchema | None = None


class NotifyQuietHoursSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    enabled: bool = True
    start: str = "22:00"
    end: str = "07:00"


class NotifySettingsSchema(BaseModel):
    """Local-notify settings blob under ``notify_settings``."""

    # ignore unknown keys; avoid OpenAPI additionalProperties index signature
    # that breaks FE assignability.
    model_config = ConfigDict(extra="ignore")

    enabled: bool = False
    voiceEnabled: bool = True
    flashEnabled: bool = True
    flashMode: Literal["timed", "persistent"] = "timed"
    leadOffsetsMinutes: list[int] = Field(default_factory=lambda: [60])
    preambleChimeId: str = "broadcast"
    quietHours: NotifyQuietHoursSchema | None = None


class NotifySettingsResponse(BaseModel):
    configured: bool
    settings: NotifySettingsSchema | None = None


class NotifySettingsBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    settings: NotifySettingsSchema


class NotifyFiredResponse(BaseModel):
    configured: bool
    keys: list[str] | None = None


class NotifyFiredBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    keys: list[str]


class NotifyFiredClaimResponse(BaseModel):
    configured: bool
    claimed: list[str] = Field(default_factory=list)
    keys: list[str] = Field(default_factory=list)


class NotifyHistoryEntrySchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    triggerReason: str
    status: NotifyHistoryStatus
    errorMessage: str | None = None
    triggeredAt: str
    eventId: str | None = None
    title: str | None = None
    leadOffsetMinutes: int | None = None


class NotifyHistoryResponse(BaseModel):
    configured: bool
    entries: list[NotifyHistoryEntrySchema] | None = None


class NotifyHistoryBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    entries: list[NotifyHistoryEntrySchema]


class AssistantToolCallSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    name: str
    resultSummary: str | None = None


class AssistantSessionMessageSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    role: AssistantMessageRole
    content: str
    toolCalls: list[AssistantToolCallSchema] | None = None


class AssistantSessionSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    title: str
    updatedAt: int
    messages: list[AssistantSessionMessageSchema]
    sessionId: str | None = None
    #: When set, assistant chat for this session uses that complete profile.
    #: When unset: hard-bound global assistant slot.
    llmProfileId: str | None = None


class AssistantSessionsResponse(BaseModel):
    configured: bool
    sessions: list[AssistantSessionSchema] | None = None
    activeSessionId: str | None = None


class AssistantSessionsPutBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    deviceId: str
    sessions: list[AssistantSessionSchema]
    activeSessionId: str | None = None


class AssistantVoiceIoSettingsSchema(BaseModel):
    """Assistant STT/TTS IO defaults (incl. calendar create target workset)."""

    # ignore unknown keys; avoid OpenAPI additionalProperties index signature.
    model_config = ConfigDict(extra="ignore")

    sttProvider: str = "browser"
    ttsProvider: str = "browser"
    ttsEnabled: bool = True
    speechLanguage: str = "zh-HK"
    spacePttMode: SpacePttMode = "hold"
    ttsVoiceUri: str = ""
    defaultWorksetId: str = SYSTEM_WORKSET_ID


class AssistantVoiceIoResponse(BaseModel):
    configured: bool
    settings: AssistantVoiceIoSettingsSchema | None = None


class AssistantVoiceIoBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    settings: AssistantVoiceIoSettingsSchema


class TimelineEventTimeOverrideSchema(BaseModel):
    model_config = ConfigDict(extra="forbid")

    startTime: str
    endTime: str | None


class TimelineAnnotationsResponse(BaseModel):
    configured: bool
    eventStatuses: dict[str, TimelineEventStatus] | None = None
    eventTimeOverrides: dict[str, TimelineEventTimeOverrideSchema] | None = None


class TimelineAnnotationsPutBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    eventStatuses: dict[str, TimelineEventStatus]
    eventTimeOverrides: dict[str, TimelineEventTimeOverrideSchema]


class ScheduleEmojisResponse(BaseModel):
    """Per-schedule-item emoji glyphs stored under ``schedule_emojis``."""

    configured: bool
    emojis: dict[str, str] | None = None


class ScheduleEmojisPutBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    emojis: dict[str, str]


class TaskEmojisResponse(BaseModel):
    """Per-analysis-task emoji glyphs stored under ``task_emojis`` (not a DB column)."""

    configured: bool
    emojis: dict[str, str] | None = None


class TaskEmojisPutBody(BaseModel):
    model_config = ConfigDict(extra="forbid")

    emojis: dict[str, str]

File: web/src/api/uiPrefs.ts
/**
 * REST client for SQLite-backed UI prefs (ops board, voice reminder, assistant…).
 *
 * INVARIANTS:
 * - After successful hydrate/PUT, server (SQLite) is SoT. Empty / unconfigured
 *   server → client defaults. Retired localStorage migration/cleanup bridges
 *   are not part of this boundary.
 * - Device chrome stays local forever: `im:monitor-mode`, `im:pages-last-path`
 *   (see `MonitorModeContext` / `boardPrefsStore`) — never put those in ui-prefs.
 * - Browse-position UI (F5 restore) stays on local `im:*` keys via
 *   `usePersistedState` — constants live next to each feature, e.g.
 *   `useTimelineNavigation` (time cursor/scale), `useTimeFilter`
 *   (`INTELLIGENCE_TIME_PRESET_STORAGE_KEY`), `mapViewHelpers` (map live/window),
 *   `systemTaskCatalog` (tasks search/mode), `*PersistedKeys.ts` on
 *   leaderboard/logs/viewer, `useTimelinePageContainer` (focused day/gantt pick).
 *   Board gantt day/month zoom is server-side only:
 *   `widgetState.ganttViewModes` on `/ui-prefs/board` (not localStorage).
 * - Unsaved drafts use sessionStorage (`storage: "session"`): chat-editor form/
 *   messages/input, `im:assistant:composer-drafts`, `im:user:profile:draft:v1`.
 * - Panel chrome (filter open / channel expand) is device-local `im:*` too
 *   (`monitorPersistedKeys`, timeline filter/event-list keys, discord expand).
 * - Timeline client statuses / time overrides: `/ui-prefs/timeline/annotations`
 *   (not soft-dismiss; that stays on `/timeline/dismissals`).
 * - UI locale remains LS-first (`i18n/locale.ts`); do not make server authoritative
 *   for `auto` locale the same way as board layout.
 * Hand-written (same pattern as timelineDismissals.ts). OpenAPI paths live under
 * `/api/v1/ui-prefs/*`; regenerate types via `npm run openapi:types` when desired.
 */

import { apiClient } from "./client";

// ── Board (layout + widgetState) ─────────────────────────────────────────

export type BoardMapViewPref = {
  center: [number, number];
  zoom: number;
};

/** Per-widget gantt day/month zoom (board gantt widgets). */
export type BoardGanttViewMode = "day" | "month";

/**
 * Per-widget source filter: hierarchical `{taskIds, worksetIds}`, or `null`
 * (all sources). Flat `string[]` is rejected on read (server sanitize may still
 * normalize older SQLite payloads before they reach the client).
 */
export type BoardSourceFilterPref =
  | { taskIds: string[]; worksetIds: string[] }
  | null;

export type BoardWidgetStatePref = {
  mapViews: Record<string, BoardMapViewPref>;
  sourceFilters: Record<string, BoardSourceFilterPref>;
  ganttViewModes: Record<string, BoardGanttViewMode>;
};

/** GET payload after server sanitize (hierarchical only). */
export type BoardWidgetStatePrefWire = BoardWidgetStatePref;

/** Layout blob stored under `ops_board_layout` (v14 widgets mosaic). */
export type BoardLayoutPref = {
  version: number;
  widgets: Array<{
    i: string;
    type: string;
    col: number;
    row: number;
    sizeId: string;
    z?: number;
  }>;
};

export type BoardPrefsResponse = {
  configured: boolean;
  layout: BoardLayoutPref | null;
  widgetState: BoardWidgetStatePrefWire | null;
};

export type BoardPrefsPutBody = {
  /** Omit to leave unchanged; `null` clears the server key. */
  layout?: BoardLayoutPref | null;
  widgetState?: BoardWidgetStatePref | null;
};

const BOARD_PATH = "/api/v1/ui-prefs/board";

/** GET `/api/v1/ui-prefs/board` — empty/missing → `{ configured: false, … null }`. */
export function fetchBoardPrefs(): Promise<BoardPrefsResponse> {
  return apiClient.get<BoardPrefsResponse>(BOARD_PATH);
}

/** PUT `/api/v1/ui-prefs/board` — partial body; returns the merged prefs. */
export function putBoardPrefs(body: BoardPrefsPutBody): Promise<BoardPrefsResponse> {
  return apiClient.put<BoardPrefsResponse>(BOARD_PATH, body);
}

// ── Voice reminder ───────────────────────────────────────────────────────

/** Settings blob stored under ``voice_reminder_settings``. */
export type VoiceReminderSettingsPayload = {
  enabled: boolean;
  leadOffsetsMinutes: number[];
  /** Hierarchical source selection (same model as board/timeline); `null` = all. */
  sourceFilter: { taskIds: string[]; worksetIds: string[] } | null;
  preambleChimeId: string;
  quietHours: { enabled: boolean; start: string; end: string };
};

export type VoiceReminderHistoryEntryPayload = {
  id: string;
  triggerReason: string;
  status: "success" | "failure";
  errorMessage: string | null;
  triggeredAt: string;
  eventId?: string;
  title?: string;
  leadOffsetMinutes?: number;
};

export type VoiceReminderSettingsResponse = {
  configured: boolean;
  settings: VoiceReminderSettingsPayload | null;
};

export type VoiceReminderFiredResponse = {
  configured: boolean;
  keys: string[] | null;
};

export type VoiceReminderFiredClaimResponse = {
  configured: boolean;
  /** Keys newly reserved for this client to speak. */
  claimed: string[];
  keys: string[];
};

export type VoiceReminderHistoryResponse = {
  configured: boolean;
  entries: VoiceReminderHistoryEntryPayload[] | null;
};

const VOICE_SETTINGS_PATH = "/api/v1/ui-prefs/voice-reminder/settings";
const VOICE_FIRED_PATH = "/api/v1/ui-prefs/voice-reminder/fired";
const VOICE_HISTORY_PATH = "/api/v1/ui-prefs/voice-reminder/history";

/** GET voice-reminder settings (`configured: false` when unset). */
export function fetchVoiceReminderSettings(): Promise<VoiceReminderSettingsResponse> {
  return apiClient.get<VoiceReminderSettingsResponse>(VOICE_SETTINGS_PATH);
}

/** PUT voice-reminder settings blob. */
export function putVoiceReminderSettings(
  settings: VoiceReminderSettingsPayload,
): Promise<VoiceReminderSettingsResponse> {
  return apiClient.put<VoiceReminderSettingsResponse>(VOICE_SETTINGS_PATH, { settings });
}

/** GET fired dedupe keys. */
export function fetchVoiceReminderFired(): Promise<VoiceReminderFiredResponse> {
  return apiClient.get<VoiceReminderFiredResponse>(VOICE_FIRED_PATH);
}

/** PUT fired dedupe keys (server prunes stale). */
export function putVoiceReminderFired(keys: string[]): Promise<VoiceReminderFiredResponse> {
  return apiClient.put<VoiceReminderFiredResponse>(VOICE_FIRED_PATH, { keys });
}

/** POST — atomically reserve keys before speak (multi-client dedupe). */
export function claimVoiceReminderFired(
  keys: string[],
): Promise<VoiceReminderFiredClaimResponse> {
  return apiClient.post<VoiceReminderFiredClaimResponse>(`${VOICE_FIRED_PATH}/claim`, {
    keys,
  });
}

/** GET trigger history entries. */
export function fetchVoiceReminderHistory(): Promise<VoiceReminderHistoryResponse> {
  return apiClient.get<VoiceReminderHistoryResponse>(VOICE_HISTORY_PATH);
}

/** PUT trigger history (server truncates to ≤100). */
export function putVoiceReminderHistory(
  entries: VoiceReminderHistoryEntryPayload[],
): Promise<VoiceReminderHistoryResponse> {
  return apiClient.put<VoiceReminderHistoryResponse>(VOICE_HISTORY_PATH, { entries });
}

// ── Assistant sessions + voice IO ─────────────────────────────────────────

export type AssistantSessionMessagePayload = {
  id: string;
  role: "user" | "assistant";
  content: string;
  toolCalls?: Array<{ name: string; resultSummary?: string }>;
};

export type AssistantSessionPayload = {
  id: string;
  title: string;
  updatedAt: number;
  messages: AssistantSessionMessagePayload[];
  sessionId?: string;
};

export type AssistantSessionsResponse = {
  configured: boolean;
  sessions: AssistantSessionPayload[] | null;
  activeSessionId: string | null;
};

export type AssistantSessionsPutBody = {
  deviceId: string;
  sessions: AssistantSessionPayload[];
  activeSessionId?: string | null;
};

export type AssistantVoiceIoSettingsPayload = {
  sttProvider: string;
  ttsProvider: string;
  ttsEnabled: boolean;
  speechLanguage: string;
  spacePttMode?: "hold" | "toggle";
  ttsVoiceUri?: string;
  /** Default assistant create target; ``__user__`` = 一般. */
  defaultWorksetId?: string;
};

export type AssistantVoiceIoResponse = {
  configured: boolean;
  settings: AssistantVoiceIoSettingsPayload | null;
};

const ASSISTANT_SESSIONS_PATH = "/api/v1/ui-prefs/assistant/sessions";
const ASSISTANT_VOICE_IO_PATH = "/api/v1/ui-prefs/assistant/voice-io";

/** GET assistant chat sessions for one device slot (`configured: false` when unset). */
export function fetchAssistantSessions(deviceId: string): Promise<AssistantSessionsResponse> {
  return apiClient.get<AssistantSessionsResponse>(ASSISTANT_SESSIONS_PATH, { deviceId });
}

/**
 * PUT full session list + active id.
 * Delete a session by omitting it from `sessions` (UI still supports delete).
 */
export function putAssistantSessions(
  body: AssistantSessionsPutBody,
): Promise<AssistantSessionsResponse> {
  return apiClient.put<AssistantSessionsResponse>(ASSISTANT_SESSIONS_PATH, body);
}

/** GET assistant STT/TTS IO settings. */
export function fetchAssistantVoiceIo(): Promise<AssistantVoiceIoResponse> {
  return apiClient.get<AssistantVoiceIoResponse>(ASSISTANT_VOICE_IO_PATH);
}

/** PUT assistant STT/TTS IO settings. */
export function putAssistantVoiceIo(
  settings: AssistantVoiceIoSettingsPayload,
): Promise<AssistantVoiceIoResponse> {
  return apiClient.put<AssistantVoiceIoResponse>(ASSISTANT_VOICE_IO_PATH, { settings });
}

// ── Timeline annotations (statuses + time overrides) ──────────────────────

export type TimelineEventStatusPayload = "pending" | "confirmed" | "completed";

export type TimelineEventTimeOverridePayload = {
  startTime: string;
  endTime: string | null;
};

export type TimelineAnnotationsPayload = {
  eventStatuses: Record<string, TimelineEventStatusPayload>;
  eventTimeOverrides: Record<string, TimelineEventTimeOverridePayload>;
};

export type TimelineAnnotationsResponse = {
  configured: boolean;
  eventStatuses: Record<string, TimelineEventStatusPayload> | null;
  eventTimeOverrides: Record<string, TimelineEventTimeOverridePayload> | null;
};

const TIMELINE_ANNOTATIONS_PATH = "/api/v1/ui-prefs/timeline/annotations";

/** GET timeline client annotations (`configured: false` when unset). */
export function fetchTimelineAnnotations(): Promise<TimelineAnnotationsResponse> {
  return apiClient.get<TimelineAnnotationsResponse>(TIMELINE_ANNOTATIONS_PATH);
}

/** PUT full timeline annotations blob (statuses + time overrides). */
export function putTimelineAnnotations(
  body: TimelineAnnotationsPayload,
): Promise<TimelineAnnotationsResponse> {
  return apiClient.put<TimelineAnnotationsResponse>(TIMELINE_ANNOTATIONS_PATH, body);
}
